Create garbage collection docs (#28870)
kubernetes/website: content/en/docs/concepts/overview/working-with-objects/finalizers.md

prevent deletion of unmanaged resources.

A common example of a finalizer is `kubernetes.io/pv-protection`, which prevents
accidental deletion of `PersistentVolume` objects.

## Owner references and finalizers

[Owner references](/concepts/overview/working-with-objects/owners-dependents/) describe the relationships between objects in Kubernetes.
Ownership is different to
[labels](/concepts/overview/working-with-objects/labels/). Kubernetes processes
finalizers when it identifies owner references on a resource targeted for
deletion. 

In some situations, finalizers can block the deletion of dependent objects,
which can cause the targeted owner object to remain in a read-only state without
being fully deleted. In these situations, you should check finalizers and owner
references on the target owner and dependent objects to troubleshoot the cause. 

{{<note>}}
In cases where objects are stuck in a deleting state, try to avoid manually
removing finalizers to allow deletion to continue. Finalizers are usually added
to resources for a reason, so forcefully removing them can lead to issues in
your cluster. 
{{</note>}}

## {{% heading "whatsnext" %}}

> * Learn about working with finalizers.

Reviewer: This should be a hyperlink, right?

Author: Yes, I haven't drafted the task topic for that yet.

---
title: Finalizers
content_type: concept
weight: 60
---

<!-- overview -->

In Kubernetes, finalizers are keys that allow you to specify pre-delete
operations for resources. You can use finalizers to control garbage collection

> of resources by alerting controllers to perform specific cleanup tasks before
> deleting the target resource. 

Reviewer: nits:
```suggestion
of resources by alerting {{< glossary_tooltip term_id="controller" text="controller">}} to
perform specific cleanup tasks before deleting the target resource.
```

Finalizers don't usually specify the code to execute. Instead, they are
typically lists of keys on a specific resource similar to annotations.
Kubernetes specifies some finalizers automatically, but you can also specify
your own.

## How finalizers work

When you create a resource using a manifest file, you can specify finalizers in
the `metadata.finalizers` field. When you attempt to delete the resource, the
controller that manages it notices the values in the `finalizers` field and does
the following: 

  * Modifies the object to add a `metadata.deletionTimestamp` field with the
    time you started the deletion.

>   * Marks the object as read-only until the `finalizers` field is empty.

Reviewer: ```suggestion
  * Marks the object as read-only until its `metadata.finalizers` field is empty.
```

The controller then attempts to satisfy the requirements of the finalizers
specified for that resource. Each time a finalizer condition is satisfied, the
controller removes that key from the resource's `finalizers` field. When the

> field is empty, garbage collection continues. You can also use finalizers to

Reviewer: ```suggestion
field is empty, {{< glossary_tooltip term_id="garbage-collection" text="garbage collection">}}
continues. You can also use finalizers to
```

Author: Ack - adding the tooltip in the first paragraph on this page.

prevent deletion of unmanaged resources.

A common example of a finalizer is `kubernetes.io/pv-protection`, which prevents
accidental deletion of `PersistentVolume` objects. When a `PersistentVolume`
object is in use by a Pod, Kubernetes adds the `pv-protection` finalizer. If you
try to delete the `PersistentVolume`, it enters a `Terminating` status, but the
controller can't delete it because the finalizer exists. When the Pod stops
using the `PersistentVolume`, Kubernetes clears the `pv-protection` finalizer,
and the controller deletes the volume.

## Owner references and finalizers

[Owner references](/concepts/overview/working-with-objects/owners-dependents/) describe the relationships between objects in Kubernetes.

> Ownership is different to
> [labels](/concepts/overview/working-with-objects/labels/). Kubernetes processes

Reviewer: ```suggestion
When {{<glossary_tooltip text="controller" term_id="controller">}} manage objects
(such as Pods), those controllers primarily track those related objects using
{{< glossary_tooltip term_id="selectors" text="selectors" >}} and
{{< glossary_tooltip text="labels" term_id="label" >}}. For example, when a Job creates one
or more Pods, the Job controller sets labels on those Pods and then tracks changes to
any Pods that have the same label.

The Job controller also marks the created Pods with an _owner reference_ pointing at the
Job those Pods belong to. If you delete the Job whilst those Pods are still running, Kubernetes
uses the owner reference to work out which Pods to clean up (and doesn't work this out
based on labels).
```

controller that manages it notices the values in the `finalizers` field and does
the following: 

  * Modifies the object to add a `metadata.deletionTimestamp` field with the
    time you started the deletion.
  * Marks the object as read-only until its `metadata.finalizers` field is empty.

The controller then attempts to satisfy the requirements of the finalizers
specified for that resource. Each time a finalizer condition is satisfied, the
controller removes that key from the resource's `finalizers` field. When the
field is empty, garbage collection continues. You can also use finalizers to
prevent deletion of unmanaged resources.

A common example of a finalizer is `kubernetes.io/pv-protection`, which prevents
accidental deletion of `PersistentVolume` objects. When a `PersistentVolume`
object is in use by a Pod, Kubernetes adds the `pv-protection` finalizer. If you
try to delete the `PersistentVolume`, it enters a `Terminating` status, but the
controller can't delete it because the finalizer exists. When the Pod stops
using the `PersistentVolume`, Kubernetes clears the `pv-protection` finalizer,
and the controller deletes the volume.

> 

Reviewer: I thought of a real-world analogy: [safety lock outs](https://www.osha.gov/control-hazardous-energy/concepts).

If you have a protective measure in place, such as an isolator, then typically each worker adds their own tag to the isolator. Convention (and, usually, some mechanism) means that each worker is responsible for their own tag. Whilst a worker could remove a colleague's tag - they have access, and probably the tools to remove it - such removal is unsound / dangerous.

Not sure if that analogy is useful. Anyway I thought I'd mention it.
